feat: Make SVGs "Penrose-editable"

lmxlucy opened the pull request:
# Description

Related issue/PR: #1129 

Provide editability of any (SVG) diagram created in Penrose.
1. SVGs exported from Penrose contain (in metadata) the entire source code used to generate the diagram (Domain, Style, and Substance), along with the seed and software version.
2. The Penrose IDE can directly import an SVG, and automatically import this metadata.

# Implementation strategy and design decisions

Penrose SVG Export:
1. Figure out where SVGs are exported
2. Programmatically retrieve the Penrose version, variation string, substance code, style code, and domain code 
3. Add this code to the exported SVG metadata

Penrose SVG Import:
1. Figure out how to upload SVG (dropzone in a new SVG tab)
2. Figure out how to load trios (focus on In-browser storage)
3. Extract trios content from SVG and load this code to trios in browser

# Examples with steps to reproduce them

Export Penrose SVG
1. Load word-cloud trios from _Roger_
2. Click _Compile_ to compile diagram from trios<img width="1440" alt="Screen Shot 2022-12-09 at 8 41 38 PM" src="https://user-images.githubusercontent.com/56939921/206822537-2b5d20ca-dfb3-4c47-8eed-1f2ccc7ebee8.png">
3. Click _SVG_ to export the diagram in svg
4. Open the exported svg in text editor to see added Penrose tag<img width="484" alt="Screen Shot 2022-12-09 at 8 44 45 PM" src="https://user-images.githubusercontent.com/56939921/206822666-9274535a-4e75-4f69-9def-dfe8eb1c70f4.png">

Import Penrose SVG
1. For comparison purpose, load walk-on-spheres from _Roger_  and compile<img width="1440" alt="Screen Shot 2022-12-09 at 8 45 33 PM" src="https://user-images.githubusercontent.com/56939921/206822709-118b0bb6-c264-4159-8b9a-79687959ca5b.png">
2. Upload the word-cloud svg downloaded above in the _svg_ tab and Compile<img width="348" alt="Screen Shot 2022-12-09 at 8 48 13 PM" src="https://user-images.githubusercontent.com/56939921/206822819-37ee3ad3-5d5d-44dc-9a84-2a29ba53154d.png">
3. Check trios content and the newly compiled diagram match the uploaded svg (the titles of trios stay as walk-on-spheres)<img width="1440" alt="Screen Shot 2022-12-09 at 8 48 49 PM" src="https://user-images.githubusercontent.com/56939921/206822842-d07f47c1-72ec-420c-85db-c9c81073918e.png">


# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new ESLint warnings
- [x] I have reviewed any generated changes to the `diagrams/` folder

wodeni commented on packages/editor/src/components/SvgUploader.tsx:
```diff
@@ -0,0 +1,86 @@
+        contents: dslElem[0].textContent ?? "".trim(),
+      });
+
+      toast.success("Sucessfully uploaded SVG to editor");
```
Since the SVG should be renderable, we can directly call `compileDiagram` after upload. Can you try using `useCompileDiagram` from `state/callbacks`?

wodeni replied:
bumping this comment

wodeni commented on packages/editor/src/components/SvgUploader.tsx:
```diff
@@ -0,0 +1,86 @@
+
+      const parser = new DOMParser();
+      const xmlDoc = parser.parseFromString(svgText, "text/xml");
+      const styElem = xmlDoc.getElementsByTagName("sty");
```
I don't see the `variation` being used here. To faithfully recreate the uploaded SVG, the `variation` needs to be the same as well. You can use `setDiagramMetadata` from `diagramMetadataSelector` to set the metadata.

wodeni replied:
Also bumping this comment

wodeni reviewed (commented):
LGTM overall! Thanks again for submitting the PR 🎉!

I only have a few minor comments that might improve the UX a little bit. I tested the new feature locally and everything seems to run very smoothly. For testing, it might be useful to start adding some tests to `editor` for this (e.g. serializing all diagrams from the registry and loading them again). This might require a bit more changes and is probably worth another PR.

lmxlucy commented on packages/core/src/renderer/Renderer.ts:
```diff
@@ -4,6 +4,7 @@
  *
  */
 
+import { sub } from "engine/AutodiffFunctions";
```
No, I will delete this

lmxlucy commented on packages/editor/src/components/DiagramPanel.tsx:
```diff
@@ -41,6 +50,76 @@ export const DownloadSVG = (
 };
 
+/**
+ * TODO: Given an SVG, program triple, and version and variation strings,
```
Yes, I will delete the TODO

wodeni commented on packages/editor/src/components/SvgUploader.tsx:
```diff
@@ -0,0 +1,94 @@
+
+      setDiagramMetadata((metadata) => ({
+        ...metadata,
+        variation: diagramMetadata.variation,
```
I think you'll need to read the variation from the SVG file instead of using the one from the current workspace.

wodeni commented on packages/editor/src/components/DiagramPanel.tsx:
```diff
@@ -41,6 +50,76 @@ export const DownloadSVG = (
+  version.insertAdjacentText("afterbegin", versionStr);
+
+  // Create <variation> tag for variation string
+  const variation = document.createElementNS(
```
(basically this field will need to be loaded in the uploader)

wodeni reviewed (changes requested):
Thanks for cleaning the diff! Small comments on variation loading below.

wodeni commented on packages/editor/src/components/SvgUploader.tsx:
```diff
@@ -0,0 +1,102 @@
+
+      setDiagramMetadata((metadata) => ({
+        ...metadata,
+        variation: variationElem[0].textContent ?? "".trim(),
```
My understanding is you wanted to `trim` the whole thing?

```suggestion
        variation: (variationElem[0].textContent ?? "").trim(),
```

wodeni commented on packages/editor/src/components/SvgUploader.tsx:
```diff
@@ -0,0 +1,102 @@
+
+      setStyle({
+        name: "SVG import",
+        contents: styElem[0].textContent ?? "".trim(),
```
same `trim` issue here

wodeni commented on packages/editor/src/components/SvgUploader.tsx:
```diff
@@ -0,0 +1,102 @@
+
+      setSubstance({
+        name: "SVG import",
+        contents: subElem[0].textContent ?? "".trim(),
```
same `trim` issue here

wodeni reviewed (changes requested):
Great! `variation` works correctly now. I think it'd be great if we auto-compile uploaded diagrams as a finishing touch. Also a small comment on calls to `trim`, not a functional issue, just want to future-proof this.

wodeni commented:
> > I think it'd be great if we auto-compile uploaded diagrams as a finishing touch.
> 
> 
> 
> I tried to auto-compile diagram after upload. I tried `compileDiagram` inside `handleChange` both inside(after current workspace is updated with uploaded trios) and after `reader.onload` but diagram was not compiled and no error was displayed. I also did not find other events appropriate for `compileDiagram`. I am not sure what the problem is. Do you have any ideas on what to do? 
> 
> 

My guess is that you did not `await` for `compileDiagram`. The function is asynchronous (for retrieving math labels and resolving images mostly) and callers need to wait for it to complete. Your `onload` will need to be async too probably.

lmxlucy commented:
> > > I think it'd be great if we auto-compile uploaded diagrams as a finishing touch.
> > 
> > 
> > I tried to auto-compile diagram after upload. I tried `compileDiagram` inside `handleChange` both inside(after current workspace is updated with uploaded trios) and after `reader.onload` but diagram was not compiled and no error was displayed. I also did not find other events appropriate for `compileDiagram`. I am not sure what the problem is. Do you have any ideas on what to do?
> 
> My guess is that you did not `await` for `compileDiagram`. The function is asynchronous (for retrieving math labels and resolving images mostly) and callers need to wait for it to complete. Your `onload` will need to be async too probably.

Thank you!! Good to go now!!

wodeni commented on packages/editor/src/components/SvgUploader.tsx:
```diff
@@ -0,0 +1,106 @@
+        contents: (dslElem[0].textContent ?? "").trim(),
+      });
+
+      await new Promise(compileDiagram);
```
```suggestion
      await compileDiagram();
```

wodeni reviewed (changes requested):
Final thing: you can simply `await` the function call. Looking forward to having this in `main`!

lmxlucy commented:
> Final thing: you can simply `await` the function call. Looking forward to having this in `main`!

Thank you!

wodeni reviewed (approved):
Wonderful 💯! Merging after CI passes.

wodeni commented:
Looks like `format` is not passing. @lmxlucy do you mind running `yarn format` and push again?

wodeni commented:
Looks like something is off with your prettier configuration. Be sure to run `yarn format` at the repo root, and do not use a global installation of prettier (I guess that's the reason why the diffs have conflicts).

wodeni commented:
Looks like we have `prettier` set at `2.2.1`, not `2.3.0`. 
https://github.com/penrose/penrose/blob/3ae696e252fb76624ebc407e6aa57d5d24c9c4a9/package.json#L37

Might wanna [reset your build](https://github.com/penrose/penrose/blob/main/CONTRIBUTING.md#refresh-build)?

lmxlucy commented:
> Looks like we have `prettier` set at `2.2.1`, not `2.3.0`.
> 
> https://github.com/penrose/penrose/blob/3ae696e252fb76624ebc407e6aa57d5d24c9c4a9/package.json#L37
> 
> Might wanna [reset your build](https://github.com/penrose/penrose/blob/main/CONTRIBUTING.md#refresh-build)?

I rebuilt and format again. But my prettier version is still 2.2.1

lmxlucy commented:
> Looks like we have `prettier` set at `2.2.1`, not `2.3.0`.
> 
> https://github.com/penrose/penrose/blob/3ae696e252fb76624ebc407e6aa57d5d24c9c4a9/package.json#L37
> 
> Might wanna [reset your build](https://github.com/penrose/penrose/blob/main/CONTRIBUTING.md#refresh-build)?

I changed my prettier version to 2.3.0 and cleaned my build. But there seem to still be conflicts

wodeni commented:
> > Looks like we have `prettier` set at `2.2.1`, not `2.3.0`.
> 
> > 
> 
> > https://github.com/penrose/penrose/blob/3ae696e252fb76624ebc407e6aa57d5d24c9c4a9/package.json#L37
> 
> > 
> 
> > Might wanna [reset your build](https://github.com/penrose/penrose/blob/main/CONTRIBUTING.md#refresh-build)?
> 
> 
> 
> I changed my prettier version to 2.3.0 and cleaned my build. But there seem to still be conflicts

To clarify: `2.2.1` is the expected version. Try downgrading or refreshing the build again?
